File: Network/Retrofit/app/src/main/java/com/example/retrofit/api/RetrofitClient.java
package com.example.retrofit.api;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitClient {
    Gson gson = new GsonBuilder().setDateFormat("YYYY-MM-dd HH:mm:ss").create();
    String BASE_URL_CURRENCY = "http://apilayer.net/";
    String BASE_URL_POSTS = "https://jsonplaceholder.typicode.com/";

    private static RetrofitClient instance = null;

    private RetrofitClient(){ }

    public static RetrofitClient getInstance(){
        if(instance == null){
            instance = new RetrofitClient();
        }
        return instance;
    }

    public Retrofit getRetrofit(){
        return new Retrofit.Builder()
                .baseUrl(BASE_URL_POSTS)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
    }

    /*Kotlin
     inline fun <refied T> getService() : T = getRetrofit().create(T::class.java)

     call : RetrofitClient().getInstance().getService<RemoteService>()
     */

}

File: Architecture/SingleActivity/app/src/main/java/com/example/singleactivity/network/RetrofitClient.kt
package com.example.singleactivity.network

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor

object RetrofitClient {
    private val BASE_URL = "http://apilayer.net/"

    private fun provideHttpLoggingInterceptor(): HttpLoggingInterceptor =
        HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY)
    private fun provideOkHttpClient( httpLoggingInterceptor: HttpLoggingInterceptor): OkHttpClient =
        OkHttpClient.Builder()
            .addInterceptor(httpLoggingInterceptor)
            .build()

    fun provideRetrofitBuilder(): Retrofit {
        return Retrofit.Builder().baseUrl(BASE_URL)
            .client(provideOkHttpClient(provideHttpLoggingInterceptor()))
            .addConverterFactory(GsonConverterFactory.create())
            .build()
    }

    inline fun <reified T> getService(): T =
        provideRetrofitBuilder().create(T::class.java)
}